File: src/android/src/com/kaottessur/pymine/level/Drone.java
package com.kaottessur.pymine.level;

import com.kaottessur.pymine.GridLocationInterface;
import com.kaottessur.pymine.Power;
import com.kaottessur.pymine.level.rating.PowerRating;

public class Drone implements GridLocationInterface {
	private int row;
	private int column;
	private boolean destroyed = false;
	
	private Minefield minefield;
	private PowerRating powerRating;
	
	public Drone(Minefield minefield, PowerRating powerRating) {
		row = 0;
		column = 0;
		
		this.minefield = minefield;
		this.powerRating = powerRating;
	}
	
	public void moveDown() {
		tryToMove(row+1, column);
	}
	
	public void moveLeft() {
		tryToMove(row, column-1);
	}
	
	public void moveRight() {
		tryToMove(row, column+1);
	}
	
	public void moveUp() {
		tryToMove(row-1, column);
	}
	
	public void tryToMove(GridSquare gridSquare) {
		GridSquare currentSquare = minefield.getGridSquareAt(row, column);
		if (minefield.getCardinallyAdjacentGridSquares(currentSquare).contains(gridSquare))
			move(gridSquare);
	}
	
	public void tryToMove(int newRow, int newColumn) {
		if (validRow(newRow) && validColumn(newColumn)) {
			GridSquare newSquare = minefield.getGridSquareAt(newRow, newColumn);
			move(newSquare);
		}
	}
	
	public void move(GridSquare gridSquare) {
		move(gridSquare.getRow(), gridSquare.getColumn());
		gridSquare.droneEntered(this);
	}
	
	public void move(int newRow, int newColumn) {
		row = newRow;
		column = newColumn;
	}
	
	public void scan() {
		powerRating.usePower(Power.SCAN_POWER);
		minefield.scan(row, column, this);
	}
	
	public void defuse() {
		powerRating.usePower(Power.DEFUSE_POWER);
		minefield.defuse(row, column, this);
	}
	
	public void defuseCarefully() {
		powerRating.usePower(Power.CAREFUL_DEFUSE_POWER);
		minefield.defuseCarefully(row, column, this);
	}
	
	public void performEMP() {
		powerRating.usePower(Power.EMP_POWER);
		minefield.performEMP(row, column, this);
	}
	
	public int getRow() {
		return row;
	}
	
	public int getColumn() {
		return column;
	}
	
	public Minefield getMinefield() {
		return minefield;
	}
	
	public boolean isDestroyed() {
		return destroyed;
	}
	
	public boolean hasPower() {
		return powerRating.getPower() > 0;
	}
	
	public int getPower() {
		return powerRating.getPower();
	}
	
	public void destroy() {
		destroyed = true;
		System.out.println("Destroyed the drone");
	}
	
	private boolean validRow(int newRow) {
		return newRow >= 0 && newRow < minefield.getRowCount();
	}
	
	private boolean validColumn(int newColumn) {
		return newColumn >= 0 && newColumn < minefield.getColumnCount();
	}
}

File: src/desktop/Level/drone.py
from power_ratings import CAREFUL_DEFUSE_POWER, DEFUSE_POWER, EMP_POWER, SCAN_POWER

from nytram.core.game_engine import TheGameEngine

class Drone:
    """ Represents the Player's Drone """
    
    def __init__(self, minefield, moveRating, powerRating):
        """ Intialize the Player's Drone """
        self.row = 0
        self.column = 0
        
        self.moveRating = moveRating
        self.powerRating = powerRating
        
        self.minefield = minefield
        self.destroyed = False
        
    def scan(self):
        """ Scan the current cell """
        self.powerRating.usePower(SCAN_POWER)
        self.minefield.scan(self.row, self.column, self)
        TheGameEngine.updateUI()
        
    def defuse(self):
        """ Defuse the current cell """
        self.powerRating.usePower(DEFUSE_POWER)
        self.minefield.defuse(self.row, self.column, self)
        TheGameEngine.updateUI()
        
    def defuseCarefully(self):
        """ Carfeully defuse a Fragile mine """
        self.powerRating.usePower(CAREFUL_DEFUSE_POWER)
        self.minefield.defuseCarefully(self.row, self.column, self)
        TheGameEngine.updateUI()
        
    def performEMP(self):
        """ Performan EMP Blast """
        self.powerRating.usePower(EMP_POWER)
        self.minefield.performEMP(self.row, self.column, self)
        TheGameEngine.updateUI()
        
    def destroy(self):
        """ Destroy the drone """
        self.destroyed = True
        
    def up(self):
        """  """
        if self.row > 0:
            self.move(-1, 0)
        
    def down(self):
        """ """
        if self.row < self.minefield.rowCount()-1:
            self.move(1, 0)
    
    def left(self):
        """ """
        if self.column > 0:
            self.move(0, -1)
    
    def right(self):
        """ """
        if self.column < self.minefield.columnCount()-1:
            self.move(0, 1)
            
    def move(self, rowMovement, columnMovement):
        """ Move the drone """
        self.row += rowMovement
        self.column += columnMovement
        self.moveRating.move()
        
        square = self.minefield.getSquare(self.row, self.column)
        square.onMove(self)
        TheGameEngine.updateUI()
        
    def hasPower(self):
        """ Returns if the drone has Power """
        return self.powerRating.power > 0